org.eclipse.scanning.points/src/org/eclipse/scanning/points/MultiStepGenerator.java

```diff
@@ -38,7 +42,56 @@ public boolean isScanPointGeneratorFactory() {
 
 	@Override
 	protected ScanPointIterator iteratorFromValidModel() {
-		return new MultiStepIterator(model);
+		this.model = model;
+
+		JythonObjectFactory<ScanPointIterator> arrayGeneratorFactory = ScanPointGeneratorFactory.JArrayGeneratorFactory();
+
+		int totalSize = 0;
+		boolean finalPosWasEnd = false;
+		List<double[]> positionArrays = new ArrayList<>(model.getStepModels().size());
+		double previousEnd = 0;
+		for (StepModel stepModel : model.getStepModels()) {
+			int size = stepModel.size();
+			double pos = stepModel.getStart();
+
+			// if the start of this model is the end of the previous one, and the end of the
+			// previous was was its final point, skip the first point
+			if (finalPosWasEnd &&
+					Math.abs(stepModel.getStart() - previousEnd) < Math.abs(stepModel.getStep() / 100)) {
+				pos = pos += stepModel.getStep();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment; "pos" already holds the assigned value along all execution paths. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)